packages/firestore/src/util/async_queue.ts

```diff
@@ -23,6 +23,18 @@ import { Code, FirestoreError } from './error';
 // tslint:disable-next-line:no-any Accept any return type from setTimeout().
 type TimerHandle = any;
 
+/**
+ * Wellknown "timer" IDs used when scheduling delayed operations on the
+ * AsyncQueue. These IDs can then be used from tests to check for the presence
+ * of operations or to run them early.
+ */
+export enum TimerId {
+  ListenStreamIdle,
+  ListenStreamConnection,
+  WriteStreamIdle,
+  WriteStreamConnection
```

Comment:
If we do that then we have to allow duplicates on the queue and it becomes ambiguous which one tests are manipulating, so I'm inclined to keep them separate. But I agree the stream constructor ends up awkward. :-/  I considered having them be abstract properties on the stream instead (so each subclass would need to define them appropriately), but I'm not sure that's actually better, and abstract doesn't port super great to Obj-C.

Reply:
This strikes me as pretty ugly too but I don't think reducing the number of states is worthwhile.

What if we fully constructed the exponential backoff and passed that in to the remote store instead of constructing it here?

Similarly can we encapsulate the idle id and duration into a "delayed runner"? That way the streams just have these dumb interfaces to delay tasks without having to distinguish how long to delay or which id to use where.

Reply:
I'm not 100% sure what you're proposing.  We could do something like:

1. Rename `TimerId` enum to `TimeoutId`
2. Add a `Timeout` tuple class (encapsulating "idle id" and "duration" as you suggested):
```
class Timeout {
  constructor(public delayMs: number, public timeoutId: TimoutId) { }
}
```
3. Change 
`asyncQueue.enqueueAfterDelay(timerId: TimerId, delayMs: number, op: () => Promise<T>)` 
to: `asyncQueue.enqueueAfterTimeout(timeout: Timeout, op: () => Promise<T>)`

4. Change PersistentStream constructor to accept an ExponentialBackoff object and Timeout object:
```
  constructor(
    private queue: AsyncQueue,
    connectionTimeout: Timeout,
    // or instead: connectionBackoff: ExponentialBackoff,
    private idleTimeout: Timeout,
    protected connection: Connection,
    private credentialsProvider: CredentialsProvider
  ) { ... }
```

I don't have strong feelings about any of this, but it doesn't reduce the number of arguments to PersistentStream's constructor.

If you're proposing we actually have a DelayedRunner class/interface that encapsulates an AsyncQueue + Timeout so we can pass it into the stream and it can do `idleTimeoutRunner.run(op)`, we _could_... but going the next step and trying to make ExponentialBackoff implement DelayedRunner as well would require surgery to move all the .reset() and .resetToMax() calls out of PersistentStream and I'm not sure it is feasible or makes sense.

And at the end of the day we're still passing in an "idleBLAH" and "connectionBLAH" to the PersistentStream constructor (where "BLAH" could be "TimerId" (today) or "Timeout" or "DelayedRunner").

So based on my understanding of your suggestion, I am not really excited...  but let me know if my understanding missed.

Reply:
Nope, you understood what I meant. And yeah, it's not much of an improvement.